checks/checkforcontent.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checks
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/ossf/scorecard/checker"
+)
+
+// CheckFilesContent downloads the tar of the repository and calls the onFileContent() function
+// shellPathFnPattern is used for https://golang.org/pkg/path/#Match
+func CheckFilesContent(checkName string, shellPathFnPattern string, c *checker.CheckRequest, onFileContent func(path string, content []byte,
+	Logf func(s string, f ...interface{})) bool) checker.CheckResult {
+	r, _, err := c.Client.Repositories.Get(c.Ctx, c.Owner, c.Repo)
+	if err != nil {
+		return checker.MakeRetryResult(checkName, err)
+	}
+	url := r.GetArchiveURL()
+	url = strings.Replace(url, "{archive_format}", "tarball/", 1)
+	url = strings.Replace(url, "{/ref}", r.GetDefaultBranch(), 1)
+
+	// Using the http.get instead of the lib httpClient because
+	// the default checker.HTTPClient caches everything in the memory and it causes oom.
+
+	//https://securego.io/docs/rules/g107.html
+	//nolint
+	resp, err := http.Get(url)
+	if err != nil {
+		return checker.MakeRetryResult(checkName, err)
+	}
+	defer resp.Body.Close()
+
+	gz, err := gzip.NewReader(resp.Body)
+	if err != nil {
+		return checker.MakeRetryResult(checkName, err)
+	}
+	tr := tar.NewReader(gz)
+	res := true
+	for {
```

Comment:
Nit: update the for loop as:

```
hdr, err := tr.Next()
for err != io.EOF {
  ...
  hdr, err := tr.Next()
}
```

Reply:
good point. I've decided against doing this because there is a series of `continue` statements in the loop; so I would need to duplicate the call to `tr.Next()` for each of them. I think calling once at the beginning of the loop is less error-prone, especially if we update the function.
Resolving for now. Please re-open if you disagree.